src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -895,33 +898,66 @@ private static IEnumerable<PSTypeName> InferTypesFromCimCommand(PseudoBindingInf
             yield return new PSTypeName(typeof(CimInstance));
         }
 
-        private IEnumerable<PSTypeName> InferTypesFromForeachCommand(PseudoBindingInfo pseudoBinding)
+        private IEnumerable<PSTypeName> InferTypesFromForeachCommand(PseudoBindingInfo pseudoBinding, CommandAst commandAst)
         {
-            AstParameterArgumentPair argument;
-            if (pseudoBinding.BoundArguments.TryGetValue("Begin", out argument))
+            var previousPipelineElement = GetPreviousPipelineCommand(commandAst);
+            if (previousPipelineElement == null)
             {
-                foreach (var type in GetInferredTypeFromScriptBlockParameter(argument))
-                {
-                    yield return type;
-                }
+                yield break;
             }
-
-            if (pseudoBinding.BoundArguments.TryGetValue("Process", out argument))
+            foreach (var t in InferTypes(previousPipelineElement))
             {
-                foreach (var type in GetInferredTypeFromScriptBlockParameter(argument))
+                var thisToRestore = _context.CurrentThisType;
+                try
                 {
-                    yield return type;
-                }
-            }
+                    _context.CurrentThisType = t;
```

Comment:
In my experience, it's probably worth it.

Personally, I would write a test covering the recursion I mentioned. If the test fails, I'm done, I know the refactoring was necessary.

If the test passes, I step through the code to convince myself my test really is testing what I thought it was so I can make an informed decision on whether it's worth refactoring or not.